redwood-flexbox-engine/src/commonMain/kotlin/app/cash/redwood/FlexboxEngine.kt

```diff
@@ -0,0 +1,2041 @@
+            getFlexItemMarginEndCross(child, false) + sumCrossSize,
+          childDimension = getFlexItemSizeCross(child, false),
+        )
+        childMainMeasureSpec = MeasureSpec.getChildMeasureSpec(
+          spec = mainMeasureSpec,
+          padding = mainPaddingStart + mainPaddingEnd +
+            getFlexItemMarginStartMain(child, false) +
+            getFlexItemMarginEndMain(child, false),
+          childDimension = childMainSize,
+        )
+        child.measure(childCrossMeasureSpec, childMainMeasureSpec)
+        updateMeasureCache(i, childCrossMeasureSpec, childMainMeasureSpec, child)
+      }
+
+      // Check the size constraint after the first measurement for the child
+      // To prevent the child's width/height violate the size constraints imposed by the
+      // {@link Node#getMinWidth()}, {@link Node#getMinHeight()},
+      // {@link Node#getMaxWidth()} and {@link Node#getMaxHeight()} attributes.
+      // E.g. When the child's layout_width is wrap_content the measured width may be
+      // less than the min width after the first measurement.
+      checkSizeConstraints(child, i)
+      if (isWrapRequired(
+          mode = mainMode,
+          maxSize = mainSize,
+          currentLength = flexLine.mainSize,
+          childLength = getViewMeasuredSizeMain(child, isMainAxisHorizontal) +
+            getFlexItemMarginStartMain(child, isMainAxisHorizontal) +
+            getFlexItemMarginEndMain(child, isMainAxisHorizontal),
+          flexItem = child,
+          flexLinesSize = flexLines.size,
+        )
+      ) {
+        if (flexLine.itemCountVisible > 0) {
+          addFlexLine(flexLines, flexLine, if (i > 0) i - 1 else 0, sumCrossSize)
+          sumCrossSize += flexLine.crossSize
+        }
+        if (isMainAxisHorizontal) {
+          if (child.height == MatchParent) {
+            // This case takes care of the corner case where the cross size of the
+            // child is affected by the just added flex line.
+            // E.g. when the child's layout_height is set to match_parent, the height
+            // of that child needs to be determined taking the total cross size used
+            // so far into account. In that case, the height of the child needs to be
+            // measured again note that we don't need to judge if the wrapping occurs
+            // because it doesn't change the size along the main axis.
+            childCrossMeasureSpec = MeasureSpec.getChildMeasureSpec(
+              spec = crossMeasureSpec,
+              padding = padding.top + padding.bottom + child.margin.top +
+                child.margin.bottom + sumCrossSize,
+              childDimension = child.height,
+            )
+            child.measure(childMainMeasureSpec, childCrossMeasureSpec)
+            checkSizeConstraints(child, i)
+          }
+        } else {
+          if (child.width == MatchParent) {
+            // This case takes care of the corner case where the cross size of the
```

Comment:
Off-topic: what kind of monster copy-pastes a 6-line comment? (This is from the original code obvs.)